### ISSUE TITLE
Liquid exception in projects.html

### ISSUE THREAD

**Issue by paddyroddy** [mcred/mcred.github.io#1]
I found your medium article very useful and have built my own website in a similar manner. In your projects.html you pull the data from GitHub and display public repositories. I find it works for me most of the time but sometimes I get the error `Liquid Exception: Liquid error (line 9): Cannot sort a null object. in projects.html` and cannot deploy my website locally or otherwise when this is the case. The line in question for me is `{% assign public_repos = site.data.github | sort: 'updated_at' | reverse %}`. After a while it works again, it appears to be random whether it works or not. Do you have the same problem?

**Comment by mcred** [comment]
I have run into this before. The issue comes from the [GitHub API Rate Limits](https://developer.github.com/v3/rate_limit/). They are IP based limits for unauthenticated calls, so Travis builds all share the same limit pool. This can happen locally too during development if you exceed the rate limit. I've fixed around this by making [authenticated calls](https://developer.github.com/v3/auth/) to the GitHub API.

**Comment by paddyroddy** [comment]
Sorry I've been trying to do this and failing. How have you made the
authenticated calls?

**Comment by paddyroddy** [comment]
as in have you made an OAuth app? I have the GitHub token in my .travis.yml

**Comment by mcred** [comment]
It comes down to how you are making the API call to populate `site.data.github`. In my case, the plugin was using the following method: 
`result = Net::HTTP.get_response(URI.parse(data['source'])).body` to get the contents.

I changed it to use an authenticated method: 
`result = RestClient::Request.execute method: :get, url: data['source'], user: ENV["GITHUB_USER"], password: ENV["GITHUB_TOKEN"]]`. 

I then used travis ENV variables to manage `GITHUB_USER` and `GITHUB_TOKEN`.

**Comment by mcred** [comment]
I found this article helpful when figuring out the basic auth:

[How to use Basic Authentication with the Ruby Rest-Client gem](https://www.krautcomputing.com/blog/2015/06/21/how-to-use-basic-authentication-with-the-ruby-rest-client-gem/)

**Comment by paddyroddy** [comment]
sorry is this in `jekyll_get.rb`? I see the similar `Net::HTTP` bit in your `jekyll_xml_source.rb`

**Comment by mcred** [comment]
Is your project Open Source? I can take a look and offer advice. I did this for my work site, not my personal. You can see the code [here](https://github.com/delphix/delphix.github.io/blob/release/_plugins/jekyll_xml_source.rb).

**Comment by paddyroddy** [comment]
ah great thanks, I'll take a look at that! Yeah, it is https://github.com/paddyroddy/paddyroddy.github.io but I'll give it a play. Thanks for the help

**Comment by mcred** [comment]
Ahh ok, I know the Jekyll_Get plugin. I stopped using it because it hadn't been updated in several years. Based on the open pull requests that haven't been merged, it also doesn't look like it's maintained either. I made my own (not trying to push it here), but the idea for authenticated calls is the same. It uses `JSON.load` to get the Github API  contents. Switch this for another HTTP client that allows for Basic Auth and you'll be set.
